core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -307,24 +341,38 @@ class LogCleaner(initialConfig: CleanerConfig,
             case _: LogCleaningAbortedException => // task can be aborted, let it go.
             case _: KafkaStorageException => // partition is already offline. let it go.
             case e: IOException =>
-              val msg = s"Failed to clean up log for ${cleanable.topicPartition} in dir ${cleanable.log.dir.getParent} due to IOException"
-              logDirFailureChannel.maybeAddOfflineLogDir(cleanable.log.dir.getParent, msg, e)
+              var logDirectory = cleanable.log.dir.getParent
+              val msg = s"Failed to clean up log for ${cleanable.topicPartition} in dir ${logDirectory} due to IOException"
+              cleanerManager.markDirectoryUncleanable(logDirectory)
+              logDirFailureChannel.maybeAddOfflineLogDir(logDirectory, msg, Some(e))
           } finally {
```

Comment:
While we're at it, is it worth adding OutOfMemoryError exceptions?  That would fix KAFKA-6533 too.

Reply:
This wouldn't fix the reported problem as far as I'm aware. KAFKA-6553 is about the client running out of memory while the LogCleaner runs in the broker.
And in the case where an OutOfMemoryError happens, we want to raise it the user. Withholding it would do not good as far as I'm aware